### ISSUE TITLE
Unable to use both GPUs 

### ISSUE THREAD

**Issue by nyxynyx** [ray-project/ray_lightning#23]
Hello, thanks for this amazing library for Lightning!

I am trying to run Lighting and Ray Tune on a system with 2 GPU. As a start, I want to use both of GPU to train 1 trial at a time.

However, when I use

```py
def train_model(config):
    ...
    trainer = pl.Trainer(
        gpus=2,
        accelerator="ddp",
        callbacks=[checkpoint_callback, tune_report_callback],
        plugins=[RayPlugin(num_workers=1, use_gpu=True)],
        precision=16,
    )
    trainer.fit(model, dm)

if __name__ == "__main__":

    ray.init()

    config = {"batch_size": 256}

    analysis = tune.run(
        train_model,
        metric="loss",
        mode="min",
        config=config,
        num_samples=1,
        resources_per_trial={"gpu": 2},
        name="test",
    )
```

I get the error about the actor or task not being able to be scheduled.

```py
== Status ==
Memory usage on this node: 17.5/125.8 GiB
Using FIFO scheduling algorithm.
Resources requested: 1/32 CPUs, 2/2 GPUs, 0.0/70.46 GiB heap, 0.0/23.58 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/x/ray_results/test
Number of trials: 1/1 (1 RUNNING)

...

(pid=109446) GPU available: True, used: True
(pid=109446) TPU available: None, using: 0 TPU cores
(pid=109446) Using native 16bit precision.
(pid=109446) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]
2021-03-30 01:12:21,546 WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffff609630d00bec4e0790a0da3f01000000 cannot be scheduled right now. It requires {CPU: 1.000000}, {GPU: 1.000000} for placement, but this node only has remaining {31.000000/32.000000 CPU, 70.556641 GiB/70.556641 GiB memory, 0.000000/2.000000 GPU, 23.583984 GiB/23.583984 GiB object_store_memory, 1.000000/1.000000 node:192.168.1.159, 1.000000/1.000000 accelerator_type:RTX}
. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

Firstly, I tried changing `Trainer` to use `gpus=1` and `num_workers=2`, but we still get the same error

```py
    trainer = pl.Trainer(
        gpus=1,
        accelerator="ddp",
        callbacks=[checkpoint_callback, tune_report_callback],
        plugins=[RayPlugin(num_workers=2, use_gpu=True)],
        precision=16,
    )
```

```py
== Status ==
Memory usage on this node: 17.1/125.8 GiB
Using FIFO scheduling algorithm.
Resources requested: 1/32 CPUs, 2/2 GPUs, 0.0/70.07 GiB heap, 0.0/23.44 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/x/ray_results/test
Number of trials: 1/1 (1 RUNNING)

...

(pid=111559) GPU available: True, used: True
(pid=111559) TPU available: None, using: 0 TPU cores
(pid=111559) Using native 16bit precision.
(pid=111559) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]
2021-03-30 01:14:17,641 WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffff98dc17f283c4962e8f312ee401000000 cannot be scheduled right now. It requires {CPU: 1.000000}, {GPU: 1.000000} for placement, but this node only has remaining {31.000000/32.000000 CPU, 70.068359 GiB/70.068359 GiB memory, 0.000000/2.000000 GPU, 1.000000/1.000000 accelerator_type:RTX, 1.000000/1.000000 node:192.168.1.159, 23.437500 GiB/23.437500 GiB object_store_memory}
. In total there are 0 pending tasks and 2 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

Next, I reduced `num_workers=2` to `num_workers=1`, and the error remains

```py
    trainer = pl.Trainer(
        val_check_interval=0.1,
        gpus=1,
        accelerator="ddp",
        callbacks=[checkpoint_callback, tune_report_callback],
        plugins=[RayPlugin(num_workers=1, use_gpu=True)],
        precision=16,
        progress_bar_refresh_rate=1000,  # refresh every 1000 iterations
    )
```

```py
== Status ==
Memory usage on this node: 19.4/125.8 GiB
Using FIFO scheduling algorithm.
Resources requested: 1/32 CPUs, 2/2 GPUs, 0.0/68.7 GiB heap, 0.0/23.05 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/x/ray_results/test
Number of trials: 1/1 (1 RUNNING)

...

(pid=113817) GPU available: True, used: True
(pid=113817) TPU available: None, using: 0 TPU cores
(pid=113817) Using native 16bit precision.
(pid=113817) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]
2021-03-30 01:16:23,959 WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffffa8b7c70cbc343023efcc60f501000000 cannot be scheduled right now. It requires {CPU: 1.000000}, {GPU: 1.000000} for placement, but this node only has remaining {31.000000/32.000000 CPU, 68.701172 GiB/68.701172 GiB memory, 0.000000/2.000000 GPU, 23.046875 GiB/23.046875 GiB object_store_memory, 1.000000/1.000000 accelerator_type:RTX, 1.000000/1.000000 node:192.168.1.159}
. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

Finally, I reduced `resources_per_trial` to `resources_per_trial={"gpu": 1},` and it finally runs, but appears to be using only 1 GPU, not 2.

```py
    trainer = pl.Trainer(
        gpus=1,
        accelerator="ddp",
        callbacks=[checkpoint_callback, tune_report_callback],
        plugins=[RayPlugin(num_workers=1, use_gpu=True)],
        precision=16,
    )
```

```py
    analysis = tune.run(
        train_model,
        metric="loss",
        mode="min",
        config=config,
        num_samples=1,
        resources_per_trial={"gpu": 1},
        name="test",
    )
```

```py
== Status ==
Memory usage on this node: 18.8/125.8 GiB
Using FIFO scheduling algorithm.
Resources requested: 1/32 CPUs, 1/2 GPUs, 0.0/69.24 GiB heap, 0.0/23.19 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/x/ray_results/test
Number of trials: 1/1 (1 RUNNING)

...

(pid=119147) GPU available: True, used: True
(pid=119147) TPU available: None, using: 0 TPU cores
(pid=119147) Using native 16bit precision.
(pid=119147) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
(pid=119167) initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/1
(pid=119167) 
(pid=119167)   | Name  | Type          | Params
(pid=119167) ----------------------------------------
(pid=119167) 0 | model | TestPLModel00 | 3.7 K 
(pid=119167) ----------------------------------------
(pid=119167) 3.7 K     Trainable params
(pid=119167) 0         Non-trainable params
(pid=119167) 3.7 K     Total params
(pid=119167) 0.015     Total estimated model params size (MB)
```

I am using
- pytorch                   1.7.0 
- pytorch-lightning         1.2.5 
- ray                       1.2.0                  
- ray-lightning             0.0.1                   

What should be the correct way to train a single trial using both GPU devices?

**Comment by import-antigravity** [comment]
I'm having the same error.

```
(pid=20595) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
(pid=20594) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [1]
2021-03-30 12:18:52,050	WARNING worker.py:1107 -- The actor or task with ID fffffffffffffffffb4b141b6a2f8397497ce22e01000000 cannot be scheduled right now. It requires {GPU: 1.000000}, {CPU: 1.000000} for placement, but this node only has remaining {18.000000/20.000000 CPU, 77.148438 GiB/77.148438 GiB memory, 0.000000/2.000000 GPU, 25.537109 GiB/25.537109 GiB object_store_memory, 1.000000/1.000000 node:192.168.212.210, 1.000000/1.000000 accelerator_type:RTX}
. In total there are 0 pending tasks and 2 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.(pid=20595) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
(pid=20594) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [1]
2021-03-30 12:18:52,050	WARNING worker.py:1107 -- The actor or task with ID fffffffffffffffffb4b141b6a2f8397497ce22e01000000 cannot be scheduled right now. It requires {GPU: 1.000000}, {CPU: 1.000000} for placement, but this node only has remaining {18.000000/20.000000 CPU, 77.148438 GiB/77.148438 GiB memory, 0.000000/2.000000 GPU, 25.537109 GiB/25.537109 GiB object_store_memory, 1.000000/1.000000 node:192.168.212.210, 1.000000/1.000000 accelerator_type:RTX}
. In total there are 0 pending tasks and 2 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

**Comment by athenawisdoms** [comment]
I am having the same error too on Ubuntu 20.04 system with 2 GPU devices. I am only able to train using 1 GPU.

**Comment by amogkam** [comment]
Hey guys, can you set this as `resources_per_trial` that you pass to `tune.run` for now:
```
resources_per_trial={
            "cpu": 1,
            "gpu": int(use_gpu)*0.01,
            "extra_cpu": num_workers,
            "extra_gpu": num_workers * int(use_gpu) - (0.01*num_samples)
        }
```

Completely agree that this is confusing, and I am working on an improvement so that resources are completely abstracted away from the user. I will update here once that is merged.

But, if you're curious as to how all of this works, when running with Tune, each Tune trial creates 1+`num_workers` processes, each of which requires GPU, but only the workers actually utilize the GPU (this is due to some Pytorch Lightning internals). So you have your driver process which is actually running your Python script which calls `tune.run`. Calling `tune.run` creates 1 process per trial which executes your training function (`train_model`) which needs to be run on GPU but doesn't actually execute any training, and each trial process creates `num_workers` additional processes to actually execute training. When passing in `cpu` and `gpu` to `resources_per_trial`, these are the resources needed for the trial process, and `extra_cpu` and `extra_gpu` are the resources needed for the additional `num_workers` processes that each trial creates.

So if you want to run 1 Tune trial leveraging 2 GPUs, you would pass this into `tune.run`:
```
resources_per_trial={
            "cpu": 1,
            "gpu": 0.01,
            "extra_cpu": 2,
            "extra_gpu": 1.99
        }
```

Hope this helps, and let me know if you guys have any other questions!

**Comment by dannychua** [comment]
@amogkam Thanks for the explanation!

I've tried re-running using the parameters you posted,

```py
trainer = pl.Trainer(
    gpus=1,
    accelerator="ddp",
    plugins=[RayPlugin(num_workers=2, use_gpu=True)],
    ...
)
```

```py
analysis = tune.run(
    ...
    resources_per_trial={"cpu": 1, "gpu": 0.01, "extra_cpu": 2, "extra_gpu": 1.99},
)
```

but I am still getting the same error

```py
== Status ==
Memory usage on this node: 19.2/125.8 GiB
Using FIFO scheduling algorithm.
Resources requested: 3/32 CPUs, 2.0/2 GPUs, 0.0/68.9 GiB heap, 0.0/23.1 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/x/ray_results/test
Number of trials: 1/1 (1 RUNNING)

...

(pid=769665) GPU available: True, used: True
(pid=769665) TPU available: None, using: 0 TPU cores
(pid=769665) Using native 16bit precision.
(pid=769665) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
2021-03-30 15:29:52,019 WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffffe19e0336690d7b0d6526332a01000000 cannot be scheduled right now. It requires {GPU: 1.000000}, {CPU: 1.000000} for placement, but this node only has remaining {30.000000/32.000000 CPU, 68.896484 GiB/68.896484 GiB memory, 0.990000/2.000000 GPU, 1.000000/1.000000 accelerator_type:RTX, 1.000000/1.000000 node:192.168.1.159, 23.095703 GiB/23.095703 GiB object_store_memory}
. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

**Comment by nyxynyx** [comment]
@amogkam Thanks for the explanation!

I've tried re-running using the parameters you posted,

```py
trainer = pl.Trainer(
    gpus=1,
    accelerator="ddp",
    plugins=[RayPlugin(num_workers=2, use_gpu=True)],
    ...
)
```

```py
analysis = tune.run(
    ...
    resources_per_trial={"cpu": 1, "gpu": 0.01, "extra_cpu": 2, "extra_gpu": 1.99},
)
```

but I am still getting the same error

```py
== Status ==
Memory usage on this node: 19.2/125.8 GiB
Using FIFO scheduling algorithm.
Resources requested: 3/32 CPUs, 2.0/2 GPUs, 0.0/68.9 GiB heap, 0.0/23.1 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/x/ray_results/test
Number of trials: 1/1 (1 RUNNING)

...

(pid=769665) GPU available: True, used: True
(pid=769665) TPU available: None, using: 0 TPU cores
(pid=769665) Using native 16bit precision.
(pid=769665) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
2021-03-30 15:29:52,019 WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffffe19e0336690d7b0d6526332a01000000 cannot be scheduled right now. It requires {GPU: 1.000000}, {CPU: 1.000000} for placement, but this node only has remaining {30.000000/32.000000 CPU, 68.896484 GiB/68.896484 GiB memory, 0.990000/2.000000 GPU, 1.000000/1.000000 accelerator_type:RTX, 1.000000/1.000000 node:192.168.1.159, 23.095703 GiB/23.095703 GiB object_store_memory}
. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

**Comment by amogkam** [comment]
@nyxynyx ah right, so there are 2 options here

You can make this change to the source code, and then re-install the package via `pip install -e .`. In https://github.com/ray-project/ray_lightning/blob/main/ray_lightning/ray_ddp.py#L98, set `num_gpus=int(use_gpu)-0.01`. I know this is very hacky, and I am currently working on improvement for this.

Or, you set `num_workers=1`, and 
```
resources_per_trial={
            "cpu": 1,
            "gpu": 1,
            "extra_cpu": 1,
            "extra_gpu": 1
        }
```
which is cleaner, but will result in 1 of the GPUs not being utilized.

**Comment by import-antigravity** [comment]
What is the purpose of the fractional gpu assignment?

**Comment by import-antigravity** [comment]
Followup question, I'm still not sure how to assign multiple CPUs to a worker without assigning multiple GPUs

**Comment by nyxynyx** [comment]
@amogkam Got both GPU utilized using your hacky fix, which I modified slightly to

    num_gpus=int(self.use_gpu)-0.01)

but as just the training was about to start (`nvidia-smi` shows the GPU memory filling up), there's a new error

> RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1603729062494/work/torch/lib/c10d/ProcessGroupNCCL.cpp:784, unhandled cuda error, NCCL version 2.7.8

Starting the Python training script using `NCCL_IB_DISABLE=1 python tune.py` does not help.

Enabled debug messages using `NCCL_DEBUG="INFO"  NCCL_IB_DISABLE=1 python tune.py`, and the following new messages appeared:

```py
(pid=909597) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1603729062494/work/torch/lib/c10d/ProcessGroupNCCL.cpp:784, unhandled cuda error, NCCL version 2.7.8
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Channel 00/02 :    0   1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Channel 01/02 :    0   1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Channel 00 : 0[2d000] -> 1[2e000] via P2P/IPC
(pid=909608) 
(pid=909608) z-pc:909608:910337 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=909608) z-pc:909608:910337 [0] NCCL INFO transport.cc:68 -> 1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO init.cc:766 -> 1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO init.cc:840 -> 1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=909611) z-pc:909611:910338 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=909611) z-pc:909611:910338 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] -1/-1/-1->1->0|0->1->-1/-1/-1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO Channel 00 : 1[2e000] -> 0[2d000] via P2P/IPC
(pid=909611) 
(pid=909611) z-pc:909611:910338 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=909611) z-pc:909611:910338 [0] NCCL INFO transport.cc:68 -> 1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO init.cc:766 -> 1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO init.cc:840 -> 1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
2021-03-30 18:20:32,548 ERROR trial_runner.py:616 -- Trial train_model_190f1_00000: Error processing event.
```

**Comment by nyxynyx** [comment]
Upgraded Pytorch to 1.8.1, a similar but different NCCL error appears

```py
ray::ImplicitFunc.train_buffered() (pid=928774, ip=192.168.1.159)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 248, in run
    self._entrypoint()
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 315, in entrypoint
    return self._trainable_func(self.config, self._status_reporter,
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 576, in _trainable_func
    output = fn()
  File "tune.py", line 54, in train_model
    trainer.fit(model, dm)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 169, in start_training
    results = process_results(futures, queue)
  File "/home/x/ray_lightning/ray_lightning/util.py", line 103, in process_results
    ray.get(ready)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
ray.exceptions.RayTaskError(RuntimeError): ray::RayExecutor.execute() (pid=928769, ip=192.168.1.159)
  File "python/ray/_raylet.pyx", line 480, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 432, in ray._raylet.execute_task.function_executor
  File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 31, in execute
    return fn(*args, **kwargs)
  File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 223, in train_remote
    super(RayPlugin, self).new_process(
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 157, in new_process
    self.configure_ddp()
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 195, in configure_ddp
    self._model = DistributedDataParallel(
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 446, in __init__
    self._sync_params_and_buffers(authoritative_rank=0)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 457, in _sync_params_and_buffers
    self._distributed_broadcast_coalesced(
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 1155, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554788289/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
ncclUnhandledCudaError: Call to CUDA function failed.
```
Also, when using `NCLL_DEBUG="INFO"`

```py
(pid=929792) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554788289/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
(pid=929792) ncclUnhandledCudaError: Call to CUDA function failed.
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Channel 00/02 :    0   1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Channel 01/02 :    0   1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Channel 00 : 0[2d000] -> 1[2e000] via P2P/IPC
(pid=929783) 
(pid=929783) z-pc:929783:930303 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=929783) z-pc:929783:930303 [0] NCCL INFO transport.cc:68 -> 1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO init.cc:766 -> 1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO init.cc:840 -> 1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=929779) z-pc:929779:930304 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=929779) z-pc:929779:930304 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] -1/-1/-1->1->0|0->1->-1/-1/-1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO Channel 00 : 1[2e000] -> 0[2d000] via P2P/IPC
(pid=929779) 
(pid=929779) z-pc:929779:930304 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=929779) z-pc:929779:930304 [0] NCCL INFO transport.cc:68 -> 1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO init.cc:766 -> 1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO init.cc:840 -> 1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
```

Packages versions being used are now
- pytorch 1.7.0
- pytorch-lightning 1.2.5
- ray 1.2.0
- cudatoolkit 10.2.89

**Comment by amogkam** [comment]
@import-antigravity the fractional GPU is to make sure that all GPUs are being utilized. The main trial process needs to be on GPU because Pytorch Lightning requires this for when the Trainer is instantiated. However, no training actually happens on this GPU. So we said it to a low value (0.01) so that the process gets placed on a GPU, but does not reserve an entire GPU which its not actually going to use. Does this make sense?

To use multiple CPUs for a worker, you can set the `num_cpus_per_worker` arg for your plugin.

**Comment by amogkam** [comment]
@nyxynyx I created a separate issue for your question: https://github.com/ray-project/ray_lightning/issues/24

**Comment by import-antigravity** [comment]
I'm still getting the same error:
```
(pid=12623) [W reducer.cpp:1050] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters. This flag results in an extra traversal of the autograd graph every iteration, which can adversely affect performance. If your model indeed never has any unused parameters, consider turning this flag off. Note that this warning may be a false positive your model has flow control causing later iterations to have unused parameters. (function operator())
2021-04-05 16:30:51,776	WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffff18115b6e4ebbabe2ff214ae501000000 cannot be scheduled right now. It requires {CPU: 3.000000}, {GPU: 1.000000} for placement, but this node only has remaining {15.000000/20.000000 CPU, 77.197266 GiB/77.197266 GiB memory, 0.980000/2.000000 GPU, 25.585938 GiB/25.585938 GiB object_store_memory, 1.000000/1.000000 node:192.168.212.206, 1.000000/1.000000 accelerator_type:RTX}
. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

And this is what my script looks like:
```python
...
tune.run(
    tune.with_parameters(
        train_fn
    ),
    name=experiment_name,
    local_dir=log_dir,
    resources_per_trial={
                "cpu": 1,
                "gpu": int(gpu)*0.01,
                "extra_cpu": 1,
                "extra_gpu": 1 * int(gpu) - (0.01*run_args['num_samples'])
            },
    **run_args
)
```

**Comment by amogkam** [comment]
Hey @import-antigravity did you also make the changes from here https://github.com/ray-project/ray_lightning/issues/23#issuecomment-810531272?

**Comment by import-antigravity** [comment]
Ah, I think I misread what you wrote. I changed my `resources_per_trial` as shown in my comment but I missed the part where that would lead to a GPU not being utilized. Currently I'm running 2 nodes with 4 GPUs total, and only 1 GPU is being used currently. I'm going to try changing the source like you suggested and see what happens.

**Comment by import-antigravity** [comment]
> Hey @import-antigravity did you also make the changes from here [#23 (comment)](https://github.com/ray-project/ray_lightning/issues/23#issuecomment-810531272)?

I did, and now I'm getting the same error, but looking at the message it says 0.99 GPUs are available and 0.99 are needed, which is strange:

```
2021-04-06 16:31:16,980	WARNING worker.py:1107 -- The actor or task with ID ffffffffffffffff5f2b5206ae396e87dfa838a701000000 cannot be scheduled right now. It requires {CPU: 3.000000}, {GPU: 0.990000} for placement, but this node only has remaining {15.000000/20.000000 CPU, 73.144531 GiB/73.144531 GiB memory, 0.990000/2.000000 GPU, 24.365234 GiB/24.365234 GiB object_store_memory, 1.000000/1.000000 node:192.168.212.206, 1.000000/1.000000 accelerator_type:RTX}
. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.
```

**Comment by import-antigravity** [comment]
Also I've double checked that my conda environment is using the modified source

**Comment by amogkam** [comment]
@import-antigravity could you share your full code and also remind me again what your node specs (# Cpus, # Gpus) are? Thanks

**Comment by import-antigravity** [comment]
> @import-antigravity could you share your full code and also remind me again what your node specs (# Cpus, # Gpus) are? Thanks

What do you mean by full code? Like the full training script?

I'm running this on a cluster with 20 CPUs and 2 GPUs on each node. I've tried with 2 nodes and 4 nodes so far, only the head node ends up being used and only 1 GPU ends up being used.

**Comment by import-antigravity** [comment]
Here's my training script, most of it relies on my external PTL code but the relevant stuff should all be here

```python
import json
import os
from importlib import import_module
from sys import argv

from pytorch_lightning import Trainer
from pytorch_lightning.loggers import TensorBoardLogger
from ray import tune
from ray.tune.integration.pytorch_lightning import TuneReportCallback
from ray.tune.suggest import Repeater
from ray_lightning import RayPlugin
from torch import cuda

from eval.embed import InceptionEmbedding
from eval.metrics import FrechetMetric, SajjadiPrecisionRecall
from models.gan import GAN
from tune.random import RandomSearcher
from util.callbacks import CGANTensorboardImageView, EpochMetricCallback, HParamReportingCallback, \
    GANTensorboardImageView

# Should be executed "python run_experiment.py [num workers] [path/to/json]"

os.environ['TUNE_DISABLE_AUTO_CALLBACK_LOGGERS'] = '1'
os.environ['SLURM_JOB_NAME'] = 'bash'

cpus_per_task = int(argv[1])
num_workers = cpus_per_task // 2
gpu = False
if cuda.device_count():
    gpu = True

content_root = '~/ganresearch'
data_dir = os.path.join(content_root, 'data')
log_dir = os.path.join(content_root, 'logs/random-slope-cifar')

config_path = argv[2]

with open(config_path) as fp:
    config = json.load(fp)

# Extract experiment name
experiment_name = config['name']

# Extract hyperparameters
hparams = config['hparams']
hp_config = {}
for hp in hparams:
    hp_config[hp['name']] = getattr(tune, hp['type'])(hp['domain'])

# Extract and setup tuning algorithm
tune_method = config['tune']
tune_args = config['tune_args']
if tune_method == 'random':
    search_alg = Repeater(RandomSearcher(hp_config), **tune_args)

# Extract and set up metrics
metric_list = config['metrics']
metrics = []
for metric_config in metric_list:
    if metric_config['type'] == 'fid':
        metric_cls = FrechetMetric
    elif metric_config['type'] == 'prd-sajjadi':
        metric_cls = SajjadiPrecisionRecall
    else:
        raise ValueError(f'{metric_config["type"]} not recognized as a metric type')

    metric_args = {
        'name': metric_config['name'],
        'embedding': metric_config['embed'],  # Will be replaced at trial runtime
        **metric_config['metric_args']
    }

    metrics.append((metric_cls, metric_args))

# Extract GAN module
class_name = config['gan_module']
ptl_path = class_name.split('.')
module = import_module(f".{'.'.join(ptl_path[:-1])}", package='models')
ptl_cls: GAN = getattr(module, ptl_path[-1])
gan_args = config['gan_args']

# Extract tune.run() args
run_args = config['run_args']

# Extract epochs, num_workers, etc.
epochs = config['epochs']
tune_metrics = config['tune_metrics']
hparam_metrics = config['hparam_metrics']
is_conditional = config['is_conditional']

print('Config parsed')

image_view_cls = CGANTensorboardImageView if is_conditional else GANTensorboardImageView


def train_fn(config):
    embeds = {}

    # Prepare embeddings and metrics
    for _, args in metrics:
        if args['embedding'] == 'inception-2048' and 'inception-2048' not in embeds.keys():
            embeds['inception-2048'] = InceptionEmbedding(2048)
        elif args['embedding'] != 'inception-2048':
            raise ValueError('Unrecognized embedding', args['embedding'])
        args['embedding'] = embeds[args['embedding']]

    tune_callback = TuneReportCallback(tune_metrics)
    metric_callback = EpochMetricCallback(metrics, interval=10)
    report_callback = HParamReportingCallback(hparam_metrics)
    logger = TensorBoardLogger(save_dir=tune.get_trial_dir(), name="", version=tune.get_trial_id(),
                               default_hp_metric=False)
    plugin = RayPlugin(num_workers=1, num_cpus_per_worker=num_workers - 1, use_gpu=gpu)

    gan = ptl_cls(data_path=data_dir, num_workers=num_workers - 1, **gan_args, **config)

    for metric in metric_callback.all_metrics:
        gan.add_module(metric.name, metric)

    trainer_args = {
        'max_epochs': epochs,
        'callbacks': [image_view_cls(n=5), tune_callback, metric_callback, report_callback],
        'progress_bar_refresh_rate': 0,
        'default_root_dir': log_dir,
        'logger': logger,
        'plugins': [plugin]
    }

    if gpu:
        trainer_args['gpus'] = 1

    t = Trainer(**trainer_args)

    t.fit(gan)


tune.run(
    tune.with_parameters(
        train_fn
    ),
    name=experiment_name,
    local_dir=log_dir,
    resources_per_trial={
                "cpu": 1,
                "gpu": int(gpu)*0.01,
                "extra_cpu": 1,
                "extra_gpu": 1 * int(gpu) - (0.01*run_args['num_samples'])
            },
    **run_args
)
```

**Comment by amogkam** [comment]
@import-antigravity I pinged you on the Pytorch Lightning Slack

**Comment by import-antigravity** [comment]
@amogkam did the new commit resolve this?

**Comment by theoad** [comment]
Hi, @amogkam. Any update on this issue ? What is the ETA of the release expected to solve the RayPlugin issue ?

**Comment by huhubrave** [comment]
Hi, @amogkam.   has this issue been solved?

**Comment by import-antigravity** [comment]
Bump

**Comment by import-antigravity** [comment]
@amogkam has this been resolved yet?

**Comment by amogkam** [comment]
Sorry for the delay here- I'm currently working on a fix. Just have to do some more testing to make sure it's fully working but will aim to get the PR up by the end of the week.

**Comment by import-antigravity** [comment]
> Sorry for the delay here- I'm currently working on a fix. Just have to do some more testing to make sure it's fully working but will aim to get the PR up by the end of the week.

Thanks, we appreciate you!!!

**Comment by Lucecpkn** [comment]
> Sorry for the delay here- I'm currently working on a fix. Just have to do some more testing to make sure it's fully working but will aim to get the PR up by the end of the week.

Thanks for your great work!   Is there any update on this issue?

**Comment by import-antigravity** [comment]
> @nyxynyx ah right, so there are 2 options here
> 
> You can make this change to the source code, and then re-install the package via `pip install -e .`. In https://github.com/ray-project/ray_lightning/blob/main/ray_lightning/ray_ddp.py#L98, set `num_gpus=int(use_gpu)-0.01`. I know this is very hacky, and I am currently working on improvement for this.
> 
> Or, you set `num_workers=1`, and
> 
> ```
> resources_per_trial={
>             "cpu": 1,
>             "gpu": 1,
>             "extra_cpu": 1,
>             "extra_gpu": 1
>         }
> ```
> 
> which is cleaner, but will result in 1 of the GPUs not being utilized.

Got the following warning:
`WARNING trial.py:342 -- Passing `extra_*` resource requirements to `resources_per_trial` is deprecated. Please use a PlacementGroupFactory object to define your resource requirements instead.`

**Comment by Lucecpkn** [comment]
Sorry, I tried and it still doesn't work for me with 2 GPUs using the updated package. 
I try to follow the guide in the Readme, by setting 
` ray_plugin = RayPlugin(num_workers=num_workers, num_cpus_per_worker=num_cpu,  use_gpu=True)` and not setting `--gpus` for traner.   
If i set `resources_per_trial=get_tune_ddp_resources(num_workers=num_workers)`, it will not use any GPU and get stuck.
If i set `resources_per_trial=get_tune_ddp_resources(num_workers=num_workers, cpus_per_worker=num_cpu,  use_gpu=True)`, it will strangely get stuck in a loop of printing the tables:

> +-----------------------------------------------+----------+-------+----------+--------------+--------+---------------+
| Trial name                                    | status   | loc   |   margin |   batch_size |     lr |   lr_cyc_size |
|-----------------------------------------------+----------+-------+----------+--------------+--------+---------------|
...

I also tried to set the `--gpus` for the Trainner, still doesn't work.

Maybe I didn't use them correctly due to my misunderstanding of the Readme. 
Could you please publish a working example / template  of using multiple GPUs with the `ray_lightning`? 
Thanks!

**Comment by amogkam** [comment]
@Lucecpkn can you also pass in `use_gpu=True` into `get_tune_ddp_resources`

**Comment by amogkam** [comment]
@Lucecpkn I'm able to successfully run the `ray_ddp_tune.py` example with GPU.

**Comment by Lucecpkn** [comment]
> @Lucecpkn can you share more of the stdout for the case where it's getting stuck in a loop?

Hi @amogkam, thank you for the reply!

I work with 4 CPUs, 2 GPUs on a single node via Slurm. 
I want to use all 2 GPUs for every trial, so I set `num_workers=1`.  The main part of my code is as below:

```
num_workers = 1
num_cpu = 4
num_gpu = 2

ray.init(num_cpus=num_cpu, num_gpus=num_gpu)


def train_tune(...):
    ...
    # ------- Train ------------
    # Init the model
    model = MyModel(...)

    # Callback for tuning hparams
    tune_callback = TuneReportCallback(
        metrics={
            "acc": "plt/val_acc",
            "f1": "plt/val_f1_weighted"
            }, 
        on="validation_end",
    )
    ray_plugin = RayPlugin(num_workers=num_workers, num_cpus_per_worker=num_cpu, use_gpu=True)

    # Init the trainer
    tb_logger = pl_loggers.TensorBoardLogger(save_dir=tune.get_trial_dir(), name="", version=".")
    trainer = Trainer.from_argparse_args(
            args, 
            max_epochs=10,
            callbacks=[tune_callback], 
            plugins=[ray_plugin],
            deterministic=True, 
            logger=tb_logger, 
            )
    
    # Train the model
    trainer.fit(model, train_loader, val_loader)


config = {...}

scheduler = ASHAScheduler(
    max_t=5,
    grace_period=1,
    reduction_factor=2,
)

reporter = CLIReporter(
    parameter_columns=[...],
    metric_columns=[...]
)

run = tune.with_parameters(train_tune, raw_args=raw_args)
    
analysis = tune.run(
    run, name="raylt_ddp_epc1", 
    metric="acc", mode="max", 
    config=config, num_samples=6, 
    scheduler=scheduler, 
    progress_reporter=reporter, 
    resources_per_trial=get_tune_ddp_resources(num_workers=num_workers, cpus_per_worker=num_cpu, use_gpu=True), 
    local_dir = os.path.join(dir_project, "..."),
)

```

The **output** shows how it gets stuck in an infinite loop:


>   2021-08-11 18:03:17.506144: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
>   Global seed set to 1234
>   Global seed set to 1234
>   2021-08-11 18:04:54,269 INFO services.py:1245 -- View the Ray dashboard at http://127.0.0.1:8265
>   2021-08-11 18:04:57,126 WARNING tune.py:495 -- Tune detects GPUs, but no trials are using GPUs. To enable trials to use GPUs, set tune.run(resources_per_trial={'gpu': 1}...) which allows Tune to expose 1 GPU to each trial. You can also override `Trainable.default_resource_request` if using the Trainable API.
>   == Status ==
>   Memory usage on this node: 26.2/187.5 GiB
>   Using AsyncHyperBand: num_stopped=0
>   Bracket: Iter 4.000: None | Iter 2.000: None | Iter 1.000: None
>   Resources requested: 0/4 CPUs, 0/2 GPUs, 0.0/64.0 GiB heap, 0.0/49.78 GiB objects (0.0/1.0 accelerator_type:V100)
>   Result logdir: /home/lucecpkn/projects/mydir/
>   Number of trials: 6/6 (6 PENDING)
>   +-----------------------------------------------+----------+-------+----------+--------------+--------+---------------+
>   | Trial name                                    | status   | loc   |   margin |   batch_size |     lr |   lr_cyc_size |
>   |-----------------------------------------------+----------+-------+----------+--------------+--------+---------------|
>   | train_tune_4ec46_00000 | PENDING  |       |        2 |           16 | 2e-05  |          1000 |
>   | train_tune_4ec46_00001 | PENDING  |       |        2 |           16 | 1e-05  |          1000 |
>   | train_tune_4ec46_00002 | PENDING  |       |       10 |           16 | 5e-05  |          2000 |
>   | train_tune_4ec46_00003 | PENDING  |       |       10 |           16 | 5e-05  |          2000 |
>   | train_tune_4ec46_00004 | PENDING  |       |       10 |           16 | 0.0001 |          2000 |
>   | train_tune_4ec46_00005 | PENDING  |       |        2 |            8 | 5e-05  |          2000 |
>   +-----------------------------------------------+----------+-------+----------+--------------+--------+---------------+
>   
>   
>   == Status ==
>   Memory usage on this node: 25.5/187.5 GiB
>   Using AsyncHyperBand: num_stopped=0
>   Bracket: Iter 4.000: None | Iter 2.000: None | Iter 1.000: None
>   Resources requested: 0/4 CPUs, 0/2 GPUs, 0.0/64.0 GiB heap, 0.0/49.78 GiB objects (0.0/1.0 accelerator_type:V100)
>   Result logdir: /home/lucecpkn/projects/mydir/
>   Number of trials: 6/6 (6 PENDING)
>   +-----------------------------------------------+----------+-------+----------+--------------+--------+---------------+
>   | Trial name                                    | status   | loc   |   margin |   batch_size |     lr |   lr_cyc_size |
>   |-----------------------------------------------+----------+-------+----------+--------------+--------+---------------|
>   | train_tune_4ec46_00000 | PENDING  |       |        2 |           16 | 2e-05  |          1000 |
>   | train_tune_4ec46_00001 | PENDING  |       |        2 |           16 | 1e-05  |          1000 |
>   | train_tune_4ec46_00002 | PENDING  |       |       10 |           16 | 5e-05  |          2000 |
>   | train_tune_4ec46_00003 | PENDING  |       |       10 |           16 | 5e-05  |          2000 |
>   | train_tune_4ec46_00004 | PENDING  |       |       10 |           16 | 0.0001 |          2000 |
>   | train_tune_4ec46_00005 | PENDING  |       |        2 |            8 | 5e-05  |          2000 |
>   +-----------------------------------------------+----------+-------+----------+--------------+--------+---------------+
>   
>   
>   ... (keeps printing the same `== Status ==` part.)

**Comment by amogkam** [comment]
Hey @Lucecpkn, yes this is because when using with Tune the driver needs to reserve 1 CPU in addition to the resources required for training. So in your example, each trial will require 5 CPUs and 2 GPUs and thus there's not enough resources in the cluster to actually run any of the trials.

Instead, I would recommend setting `cpus_per_worker` to 1. It only makes sense to set it to more than 1 if the worker actually uses multiple CPUs (i.e. your LightningModule has to do multithreading operations). Instead you should use multiple workers to actually get parallelism. So a recommended setup would be num_workers=2, use_gpu=True and leave everything else as default. Then this should work for you.

**Comment by Lucecpkn** [comment]
Hi @amogkam,  I followed your suggestion to decrease the `cpus_per_worker` and indeed the trials can be run now. 
But I do want to keep using multiple GPUs for every trial because I want the large effective batch size which is impossible to fit on 1 GPU.   However, the Tune still uses only 1 GPU, weather I set `use_gpu`=True or 2. Any idea why?
